### PR TITLE
Test CPU metric values greater than zero

### PR DESCRIPTION
**Pull request by ChenaLee**
# Description of the issue
Based on our testing decisions, we need to add benchmark testing for metric values. To create a foundation for that, we are starting with adding sanity tests to verify we get values and they are greater than zero for all metrics.

# Description of changes
- Added CPU metric value sanity test and query logic, with extensibility for future tests in mind.

## Out of scope / next steps
- Note that some metrics weren't reporting with this set up, so this test currently fails (see test result below). Need to verify with subject matter experts in the team whether who need. Might need to change validation logic or figure out what to change in the testing environment in order to have those metrics get reported for testing purposes.
- Will refactor the main test function and structure in a way that improves readability and even easier to add new sanity and benchmark tests.
- Will add some readme or resources that helps people learn how to add tests

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Screen Shot 2022-10-11 at 14 10 56

# Requirements
I did run and committed linter changers just for the files I added or modified.
1. Run `make fmt` and `make fmt-sh`
2. Run `make linter`

### CODE REVIEW

File: integration/test/metric/metric_value_query.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+type MetricValueFetcher interface {
+	Fetch(namespace string, metricName string, stat Statistics) ([]float64, error)
+	fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error)
+	isApplicable(metricName string) bool
+	getMetricSpecificDimensions() []types.Dimension
+}
+
+type baseMetricValueFetcher struct{}
+
+func (f *baseMetricValueFetcher) fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error) {
```
Comment (SaxyPandaBear) [inline integration/test/metric/metric_value_query.go:44]: Some linters catch this, but this function doesn't actually reference `f` here, which makes me think that this function does not need to be directly tied to the MetricValueFetcher. What it looks like to me is that an instance of a MetricValueFetcher defines a set of dimensions that it allows for querying, and from there they all will make the same structured query with their specific dimensions. 

I think that this might be worthwhile to restructure, to pull out this `fetch` function. Maybe we don't need the `Fetch()` function on the fetcher interface, and just define `getMetricSpecificDimensions()`, and call this separate utility (sorry) function with that as a parameter value.

```go
cpuFetcher := CPUMetricValueFetcher{}

values, err := MyFetchUtilFunction(namespace, cpuFetcher.getDimensions(), metricName, stat)
```

File: integration/test/metric/stat.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+type Statistics int
+
+const (
+	AVERAGE Statistics = iota
+)
+
+func (s Statistics) String() string {
+	return [...]string{"Average"}[s]
+}
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric/stat.go:11]: oh nice, I tried to find something like that and couldn't at least with my quick google searches of golang :D Thanks!

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		}
+	})
+
+	// TODO: Get CPU value > 0
+	// TODO: Range test with >0 and <100
+	// TODO: Range test: which metric to get? api reference check. should I get average or test every single datapoint for 10 minutes? (and if 90%> of them are in range, we are good)
+}
+
+func validateCpuMetrics() (bool, map[string]bool) {
+	metricsToFetch := []string{
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:61]: agreed. pulled everything else out but this somehow

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			validationResult[metricName] = false
+			continue
+		}
+		if !isAllValuesGreaterThanZero(metricName, values) {
+			isAllValid = false
+			validationResult[metricName] = false
+			continue
+		}
+	}
+	return isAllValid, validationResult
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:87]: oh no! I meant to change this part of the code and completely forgot haha. Will take a look

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	log.Printf("==============%v==============", testSuite)
+	log.Printf("==============%v==============", status)
+	w := tabwriter.NewWriter(log.Writer(), 1, 1, 1, ' ', 0)
+	for metricName, isSucessful := range testSummary {
+		var status string
+		if isSucessful {
+			status = "Success"
+		} else {
+			status = "Failed"
+		}
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:124]: In the map? Could you give me more details? If it's nicer I will do it

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		fmt.Fprintln(w, metricName, "\t", status, "\t")
+	}
+	w.Flush()
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:127]: Will check and get back on this

File: integration/test/metric/query-json.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+[
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric/query-json.json:1]: no.. didn't mean to. This is to be added for my later guidelines/help material for future devs working on tests. I used it for my local cli call test to reduce iteration time

File: integration/test/metric/metric_value_query.go (hunk 1 above)
Comment (ChenaLee) [reply integration/test/metric/metric_value_query.go:44]: ew

Comment (ChenaLee) [reply integration/test/metric/metric_value_query.go:44]: jk, so I was trying to figure out how to do in golang what I'd do in java for this logic. And yeah maybe in golang world we don't need this as a... struct member function ? (I understood this as similar to what java class member functions are). I still do like having a single struct that you can call Fetch on instead of getting cpu fetcher using a factory, then passing a function in there into another util method. That seems like too much extra implementation detail to be a good interface for future tests. Let's continue to discuss what each of us care about in this code and how to satisfy both?

File: integration/test/metric/cpu.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	return exists
+}
+
+func (f *CPUMetricValueFetcher) getMetricSpecificDimensions() []types.Dimension {
+	cpuDimension := types.Dimension{
+		Name:  aws.String("cpu"),
+		Value: aws.String("cpu-total"),
+	}
+	dimensions := make([]types.Dimension, 1)
+	dimensions[0] = cpuDimension
```
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/metric/cpu.go:69]: I would have to look through this code base cause I don't remember where it is, but there is an interface defined, and then rather than implementing a function in the struct, it just defines an attribute in the struct and that satisfies it.

```go
type Foo interface {
  func Something() string
}

type Bar struct {
  Something string
}
```

[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/metric/cpu.go:69]: But that's probably not really an improvement here, that was really just an aside.

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:124]: I was thinking rather than setting bools in the map, like `m[key] = false`, you could just do `m[key] = "success"`... but maybe that's not worth it

File: integration/test/metric/cpu.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	values, err := f.fetch(namespace, dimensions, metricName, stat)
+	if err != nil {
+		log.Printf("Error while fetching metric value for %v: %v", metricName, err.Error())
+	}
+	return values, err
+}
+
+func (f *CPUMetricValueFetcher) fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error) {
+	return f.base.fetch(namespace, metricSpecificDimensions, metricName, stat)
+}
```
Comment (adam-mateen) [inline integration/test/metric/cpu.go:31]: I don't see the point of overriding the base class implementation here.

I think it would be simpler just to have:
```
type CPUMetricValueFetcher struct {
    baseMetricValueFetcher
}
```
Then you do not need to override.
You can just remove:
```
func (f *CPUMetricValueFetcher) fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error) {
	return f.base.fetch(namespace, metricSpecificDimensions, metricName, stat)
}
```
And any call to `CPUMetricValueFetcher.fetch()` will call the base class's implementation.

more details - https://golangbot.com/inheritance/

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:87]: oh yeah I see what you mean. I was hoping to find something like java map reduce stuff for this kind of thing instead of for loops but I haven't really got there. Gonna see if I can find it

[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:87]: Yeah, Golang doesn't really have those niceties yet. I think Go 1.19 introduces generics so that's a start. I wish they had more functional programming stuff but it's not there yet.

File: integration/test/metric/cpu.go (hunk 9 above)
Comment (SaxyPandaBear) [reply integration/test/metric/cpu.go:31]: I think the issue is the pointer reference in the struct, but I'm not entirely sure. I haven't wrapped my head fully around inheritance in Go yet

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:87]: so ugly :(

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:127]: nope doesn't seem like it

File: integration/test/metric/cpu.go (hunk 9 above)
Comment (ChenaLee) [reply integration/test/metric/cpu.go:31]: yeah I mean this is technically not inheritance. They consider this composition and while I'm loving that go is forcing me to always use composition rather than inheritance, I don't know how this would behave(even if it works) if I take in two structs that both implement the same method e.g. baseMetricFetcher.fetch() and complexMetricFetcher.fetch() both available

File: integration/test/metric/stat.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric/stat.go]: Done, letting test run https://github.com/ChenaLee/amazon-cloudwatch-agent/actions/runs/3237511296

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Done, letting test run https://github.com/ChenaLee/amazon-cloudwatch-agent/actions/runs/3237511296

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Done, letting test run https://github.com/ChenaLee/amazon-cloudwatch-agent/actions/runs/3237511296

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Done, letting test run https://github.com/ChenaLee/amazon-cloudwatch-agent/actions/runs/3237511296

File: integration/test/metric/cpu.go (hunk 9 above)
Comment (ChenaLee) [reply integration/test/metric/cpu.go:31]: ok this built.. I think I wasn't understanding the syntax yesterday or something. Waiting for the test to run

File: integration/test/metric/metric_value_query.go (hunk 1 above)
Comment (ChenaLee) [reply integration/test/metric/metric_value_query.go:43]: If the alternative in the below works, is it ok if we leave it this ?

File: integration/test/metric/stat.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric/stat.go]: Verified result!

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Verified result!

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Verified result!

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go]: Verified result!

File: integration/test/metric/cpu.go (hunk 9 above)
Comment (ChenaLee) [reply integration/test/metric/cpu.go]: > I don't know how this would behave(even if it works) if I take in two structs that both implement the same method
I know what happens now. This was the problem that I had to fix yesterday.

Verified things work with the tests after the change

File: integration/test/metric/metric_value_query.go (hunk 1 above)
Comment (SaxyPandaBear) [reply integration/test/metric/metric_value_query.go:43]: Adam's comment and suggestion satisfies my concern with this.

File: integration/test/metric/cpu.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	return exists
+}
+
+func (f *CPUMetricValueFetcher) getMetricSpecificDimensions() []types.Dimension {
+	cpuDimension := types.Dimension{
+		Name:  aws.String("cpu"),
+		Value: aws.String("cpu-total"),
+	}
+	dimensions := make([]types.Dimension, 1)
+	dimensions[0] = cpuDimension
```
Comment (SaxyPandaBear) [inline integration/test/metric/cpu.go:65]: This can still be simplified. I resolved the other comment cause I think I derailed it, but this could be made into a top level var, or at least just
```go
return []types.Dimension{
    {
        Name: aws.String("cpu"),
        Value: aws.String("cpu-total"),
    },
}
```

File: integration/test/metric/metric_value_query.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func GetMetricFetcher(metricName string) (MetricValueFetcher, error) {
+	for _, fetcher := range metricValueFetchers {
+		if fetcher.isApplicable(metricName) {
+			return fetcher, nil
+		}
+	}
+	err := fmt.Errorf("No metric fetcher for metricName %v", metricName)
+	log.Printf("%s", err)
```
Comment (SaxyPandaBear) [inline integration/test/metric/metric_value_query.go:30]: Same sentiment here as above. I don't think that we need to log the error here if the caller function handles it.

File: integration/test/metric/metric_value_query.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	metricQuery := types.MetricDataQuery{
+		MetricStat: &types.MetricStat{
+			Metric: &metricToFetch,
+			Period: &metricQueryPeriod,
+			Stat:   aws.String(string(stat)),
+		},
+		Id: aws.String(metricName),
+	}
+	metricDataQueries := make([]types.MetricDataQuery, 1)
+	metricDataQueries[0] = metricQuery
```
Comment (SaxyPandaBear) [inline integration/test/metric/metric_value_query.go:66]: Similarly, can simplify this so we don't `make` an array and then set a value in it, we can just define the slice literal, like the above example for the CPU dimensions comment.

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	log.Printf("==============================")
+}
+
+func getTestSuiteStatus(testSummary map[string]status.TestStatus) status.TestStatus {
+	isAllSuccessful := status.SUCCESSFUL
+	for _, value := range testSummary {
+		if value == status.FAILED {
+			isAllSuccessful = status.FAILED
+			break
```
Comment (SaxyPandaBear) [inline integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:126]: Honestly, we could just `return status.FAILED` here.

File: integration/test/metric/cpu.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+type CPUMetricValueFetcher struct {
+	baseMetricValueFetcher
+}
+
+func (f *CPUMetricValueFetcher) Fetch(namespace string, metricName string, stat Statistics) ([]float64, error) {
+	dimensions := f.getMetricSpecificDimensions()
+	values, err := f.fetch(namespace, dimensions, metricName, stat)
+	if err != nil {
+		log.Printf("Error while fetching metric value for %v: %v", metricName, err.Error())
+	}
```
[the remark this answers is not in the document]
Comment (ChenaLee) [reply integration/test/metric/cpu.go:25]: I moved this here because I hated how all those logging at callers gets distracting for readers. While I do generally hate it, I also know some people prefer to log in caller.

File: integration/test/metric/cpu.go (hunk 10 above)
Comment (ChenaLee) [reply integration/test/metric/cpu.go:65]: agreed. I think I missed that part

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 13 above)
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:126]: that is... true.. what did I do 🙈

File: integration/test/metric/metric_value_query.go (hunk 11 above)
Comment (ChenaLee) [reply integration/test/metric/metric_value_query.go:30]: ditto

File: integration/test/metric/cpu.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply integration/test/metric/cpu.go:25]: I'm fine with leaving it then

**Review by SaxyPandaBear — APPROVED** [review]
lgtm

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/aws/amazon-cloudwatch-agent/integration/test/status"
+)
+
+const configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
+const configJSON = "/base_config.json"
+
+const namespace = "MetricValueBenchmarkTest"
+const instanceId = "InstanceId"
+
+const minimumAgentRuntime = 3 * time.Minute
```
Comment (adam-mateen) [inline integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:27]: Why 3 minutes? The explanation could go in a code comment.

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+	"testing"
+	"text/tabwriter"
+	"time"
+
+	"github.com/aws/amazon-cloudwatch-agent/integration/test"
+	"github.com/aws/amazon-cloudwatch-agent/integration/test/metric"
+	"github.com/aws/amazon-cloudwatch-agent/integration/test/status"
+)
+
+const configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
```
Comment (adam-mateen) [inline integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:21]: style:

I am sort of surprised go formatter didn't automatically do this:

```
const (
    configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
    const configJSON = "/base_config.json"
    namespace = "MetricValueBenchmarkTest" instanceId = "InstanceId"
    minimumAgentRuntime = 3 * time.Minute
)
```

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+const configJSON = "/base_config.json"
+
+const namespace = "MetricValueBenchmarkTest"
+const instanceId = "InstanceId"
+
+const minimumAgentRuntime = 3 * time.Minute
+
+func TestCPUValue(t *testing.T) {
+	log.Printf("testing cpu value...")
```
Comment (adam-mateen) [inline integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:30]: nitpick: Use `Println()` instead of `Printf()` and save the planet.

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+	t.Run(fmt.Sprintf("resource file location %s ", resourcePath), func(t *testing.T) {
+		test.CopyFile(resourcePath+configJSON, configOutputPath)
+		err := test.StartAgent(configOutputPath, false)
+
+		if err != nil {
+			t.Fatalf("Agent could not start")
+		}
+
+		time.Sleep(minimumAgentRuntime)
```
Comment (adam-mateen) [inline integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:44]: It seems you are sleeping a fixed amount of time and then asserting you can fetch metrics.
When there is only a few test cases and we don't care about test time this is fine.

But as the number of test cases grow it becomes more beneficial to poll instead of waiting a fixed amount of time.

I've definitely written tests like this, so no need to fix it now.
Just something to consider in the future.

**Review by adam-mateen — APPROVED** [review]

File: integration/test/metric_value_benchmark/metrics_value_benchmark_test.go (hunk 15 above)
Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:27]: I think the variable name should be sufficient? It was minimum I had to run it for

Comment (ChenaLee) [reply integration/test/metric_value_benchmark/metrics_value_benchmark_test.go:27]: ( I don't like comments if I can avoid)
